torch/_dynamo/eval_frame.py

```diff
@@ -691,138 +696,174 @@ def result_capturing_wrapper(*graph_inputs):
+            assert index not in tensor_to_indices[tensor], f"Duplicate index {index}"
+            tensor_to_indices[tensor].add(index)
+            torch._dynamo.mark_dynamic(tensor, index)
+
+        seen_tensors = set(tensor_to_indices.keys())
+        for arg in flat_args:
+            if isinstance(arg, torch.Tensor) and arg in seen_tensors:
+                seen_tensors.remove(arg)
+        assert len(seen_tensors) == 0, "User specified dynamic dim for unknown tensors"
+
+        remove_from_cache(f)
+        with patch(f"{__name__}.most_recent_backend", None), config.patch(
+            specialize_int=True
+        ):
+            opt_f = optimize_assert(
+                dynamo_normalization_capturing_compiler,
+                hooks=Hooks(guard_export_fn=guard_export_print, guard_fail_fn=None),
+                export=True,
+                dynamic=(tracing_mode == "symbolic"),
+            )(f)
+            # TODO(voz): We may have instances of `f` that mutate inputs, we should track sideffects and reject.
+            result_traced = opt_f(*args, **kwargs)
+        remove_from_cache(f)
 
-    assert (
-        graph is not None
-    ), "Failed to produce a graph during tracing. Tracing through 'f' must produce a single graph."
-    assert out_guards is not None, "Failed to produce guards during tracing"
+        assert (
+            graph is not None
+        ), "Failed to produce a graph during tracing. Tracing through 'f' must produce a single graph."
+        assert out_guards is not None, "Failed to produce guards during tracing"
 
-    matched_input_elements_positions = produce_matching(flat_args, graph_captured_input)
+        matched_input_elements_positions = produce_matching(
+            flat_args, graph_captured_input
+        )
 
-    flat_results_traced, out_spec_traced = pytree.tree_flatten(result_traced)
+        flat_results_traced, out_spec_traced = pytree.tree_flatten(result_traced)
 
-    assert graph_captured_result is not None
-    flat_both = list(graph_captured_result) + flat_args
-    matched_output_elements_positions = produce_matching(flat_both, flat_results_traced)
+        assert graph_captured_result is not None
+        flat_both = list(graph_captured_result) + flat_args
+        matched_output_elements_positions = produce_matching(
+            flat_both, flat_results_traced
+        )
 
-    class ChangeInputOutputSignature(torch.fx.interpreter.Transformer):
-        def __init__(
-            self,
-            m,
-        ):
-            super().__init__(m)
-            arg_len = len(flat_args)
-            self.new_args = [
-                super(ChangeInputOutputSignature, self).placeholder(f"arg{i}", (), {})
-                for i in range(0, arg_len)
-            ]
-            self.old_args_gen = (
-                self.new_args[i] for i in matched_input_elements_positions
-            )
+        class ChangeInputOutputSignature(torch.fx.interpreter.Transformer):
```

Comment:
NIT: Can we move this out of export function?

Reply:
Not in this pr